File: test-suite/src/test/java/io/micronaut/docs/aop/lifecycle/ProductService.java
package io.micronaut.docs.aop.lifecycle;

// tag::imports[]
import jakarta.inject.Singleton;
import java.util.*;
// end::imports[]

// tag::class[]
@Singleton
public class ProductService {
    private final Map<String, Product> products = new HashMap<>();

    void addProduct(Product product) {
        products.put(product.getProductName(), product);
    }

    void removeProduct(Product product) {
        product.setActive(false);
        products.remove(product.getProductName());
    }

    Optional<Product> findProduct(String name) {
        return Optional.ofNullable(products.get(name));
    }
}
// end::class[]

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/aop/lifecycle/ProductService.kt
package io.micronaut.docs.aop.lifecycle

// tag::imports[]
import java.util.*
import jakarta.inject.Singleton
// end::imports[]

// tag::class[]
@Singleton
class ProductService {
    private val products: MutableMap<String, Product> = HashMap()
    fun addProduct(product: Product) {
        products[product.productName] = product
    }

    fun removeProduct(product: Product) {
        product.active = false
        products.remove(product.productName)
    }

    fun findProduct(name: String): Optional<Product> {
        return Optional.ofNullable(products[name])
    }
}
// end::class[]
